onnxruntime/core/providers/webgpu/tensor/tile.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/inlined_containers.h"
+#include "core/providers/webgpu/tensor/tile.h"
+#include "core/providers/cpu/tensor/utils.h"
+#include "core/providers/webgpu/shader_helper.h"
+#include "core/providers/webgpu/webgpu_supported_types.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(
+    Tile,
+    kOnnxDomain,
+    6, 12,
+    kWebGpuExecutionProvider,
+    (*KernelDefBuilder::Create()).TypeConstraint("T", WebGpuSupportedNumberTypes()),
+    Tile);
+
+ONNX_OPERATOR_KERNEL_EX(
+    Tile,
+    kOnnxDomain,
+    13,
+    kWebGpuExecutionProvider,
+    (*KernelDefBuilder::Create()).TypeConstraint("T", WebGpuSupportedNumberTypes()),
+    Tile);
+
+const std::string AppendTileFunction(ShaderVariableHelper& input, ShaderVariableHelper& output) {
+  std::ostringstream ss;
+  ss.imbue(std::locale::classic());
+  const TensorShape& input_shape = input.GetShape();
+  int32_t input_rank = gsl::narrow_cast<int32_t>(input_shape.NumDimensions());
+  ss << "fn tile(i: output_indices_t)->input_indices_t {\n"
+        "  var input_indices;\n";
+  for (auto i = 0; i < input_rank; i++) {
+    ss << "  input_dim_i = input.GetDimensionByIndex(" << i << ");\n";
+    ss << "  input_dim_value = output.GetDimensionByIndex(" << i << ") % input_dim_i;\n";
+    ss << "  input.indicesSet('input_indices', '" << i << "', 'input_dim_value');\n";
+  }
+  ss << "  return input_indices;\n"
+        "}\n";
+
+  return ss.str();
+}
+
+Status TileProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const ShaderVariableHelper& input = shader.AddInput("input", ShaderUsage::UseUniform);
+  const ShaderVariableHelper& output = shader.AddOutput("output", ShaderUsage::UseUniform);
+  shader.AppendImplementation(AppendTileFunction(input, output));
+  shader.SetMainFunctionBody(shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size"),
+                             "  let output_indices = ", output.OffsetToIndices("global_idx"),
+                             ";\n"
+                             "  let input_indices = tile(input, output); \n"
+                             "  ",
+                             output.SetByOffset("global_idx", input.GetByIndices("input_indices")));
+
+  return Status::OK();
+}
+
+Status Tile::ComputeInternal(ComputeContext& context) const {
+  const auto* input_tensor = context.Input(0);
+  const TensorShape& input_shape = input_tensor->Shape();
+  int32_t input_rank = gsl::narrow_cast<int32_t>(input_shape.NumDimensions());
+
+  const auto* repeats_tensor = context.Input(1);
+  const auto* repeats = repeats_tensor->Data<int32_t>();
```

Comment:
repeats_tensor is defined in spec as int64. `repeats_tensor->Data<int32_t>()` may get unexpected data.